R/oc_process.R

```diff
@@ -246,18 +245,38 @@ oc_build_url <- function(query_par, endpoint) {
 
 #' GET request from OpenCage
 #'
-#' @param oc_url character string URL with query parameters, built with
+#' @param oc_url_parts list with path and query, built with
 #'   `oc_build_url()`
 #'
-#' @return crul::HttpResponse object
+#' @return httr2 response
 #' @noRd
 
-oc_get <- function(oc_url) {
-  client <- crul::HttpClient$new(
-    url = oc_url,
-    headers = list(`User-Agent` = oc_ua_string)
-  )
-  client$get()
+oc_get <- function(oc_url_parts = NULL) {
+
+  query_req <- build_query_with_req(oc_url_parts)
+
+  query_req %>%
+    httr2::req_throttle(rate = getOption("oc_rate_sec", default = 1L) / 1L) %>%
+    httr2::req_user_agent(oc_ua_string) %>%
+    httr2::req_perform() # will error if API error :-)
+}
+
+build_query_with_req <- function(oc_url_parts) {
+  initial_req <- httr2::request("https://api.opencagedata.com")
+
+  req_with_url <- if (!is.null(oc_url_parts[["path"]])) {
+    httr2::req_url_path_append(initial_req, oc_url_parts[["path"]])
+  } else {
+    initial_req
+  }
+
+  query_req <- if (!is.null(oc_url_parts[["query"]])) {
+    httr2::req_url_query(req_with_url, !!!oc_url_parts[["query"]])
+  } else {
+    req_with_url
+  }
+
+  query_req
```

Comment:
If we only use this within `oc_process()` and tie it to the OpenCage API domain, I think the following should be just as safe. I.e. I cannot think of a scenario where we would make queries to api.opencagedata.com without an `endpoint` or `query_parameters`.

```suggestion
build_request <- function(endpoint, query_parameters) {
  httr2::request("https://api.opencagedata.com") %>%
    httr2::req_url_path_append(endpoint) %>%
    httr2::req_url_query(req_with_url, !!!query_parameters)
```

But if we want to keep it generic (e.g. for the `oc_get_memoise()` tests), we could instead do

```suggestion
build_request <- function(base_url = "https://api.opencagedata.com", endpoint, query_parameters) {
  query_req <- httr2::request(base_url)

  query_req <- if (!is.null(endpoint)) {
    httr2::req_url_path_append(query_req , endpoint)
  }

  query_req <- if (!is.null(query_parameters)) {
    httr2::req_url_query(query_req, !!!query_parameters)
  }
  
  query_req
```